<a href="https://colab.research.google.com/github/Ranjitkumarmuduli/Capstone-Project--4/blob/main/Online_Retail_Customer_Segmentation_Capstone_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## <b> Problem Description </b>

### In this project, your task is to identify major customer segments on a transnational data set which contains all the transactions occurring between 01/12/2010 and 09/12/2011 for a UK-based and registered non-store online retail.The company mainly sells unique all-occasion gifts. Many customers of the company are wholesalers.

## <b> Data Description </b>

### <b>Attribute Information: </b>

* ### InvoiceNo: Invoice number. Nominal, a 6-digit integral number uniquely assigned to each transaction. If this code starts with letter 'c', it indicates a cancellation.
* ### StockCode: Product (item) code. Nominal, a 5-digit integral number uniquely assigned to each distinct product.
* ### Description: Product (item) name. Nominal.
* ### Quantity: The quantities of each product (item) per transaction. Numeric.
* ### InvoiceDate: Invice Date and time. Numeric, the day and time when each transaction was generated.
* ### UnitPrice: Unit price. Numeric, Product price per unit in sterling.
* ### CustomerID: Customer number. Nominal, a 5-digit integral number uniquely assigned to each customer.
* ### Country: Country name. Nominal, the name of the country where each customer resides.

#Importing Some Libraries and Loading Data.


In [5]:
#Importing Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

import warnings
warnings.filterwarnings('ignore')

In [6]:
#mount google drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [7]:
#Data file path
path = '/content/drive/MyDrive/ Capstone project -4/Online Retail.xlsx - Online Retail.csv'

In [8]:
#Reading the dataset
data_1=pd.read_csv(path)

In [9]:
#copy of the dataset
df=data_1.copy()

# Dataset Exploring abd Preprocessing

In [10]:
#Shape of the dataset
df.shape

(541909, 8)

In [11]:
#First five rows of the dataset
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

    InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/10 8:26       2.55     17850.0  United Kingdom  
1  12/1/10 8:26       3.39     17850.0  United Kingdom  
2  12/1/10 8:26       2.75     17850.0  United Kingdom  
3  12/1/10 8:26       3.39     17850.0  United Kingdom  
4  12/1/10 8:26       3.39     17850.0  United Kingdom

In [12]:
#Last five rows of the dataset
df.tail()

InvoiceNo StockCode                      Description  Quantity  \
541904    581587     22613      PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899     CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254    CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255  CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138    BAKING SET 9 PIECE RETROSPOT          3   

          InvoiceDate  UnitPrice  CustomerID Country  
541904  12/9/11 12:50       0.85     12680.0  France  
541905  12/9/11 12:50       2.10     12680.0  France  
541906  12/9/11 12:50       4.15     12680.0  France  
541907  12/9/11 12:50       4.15     12680.0  France  
541908  12/9/11 12:50       4.95     12680.0  France

In [13]:
#List of the Columns name
list(df.columns)

['InvoiceNo',
 'StockCode',
 'Description',
 'Quantity',
 'InvoiceDate',
 'UnitPrice',
 'CustomerID',
 'Country']

In [14]:
#Info of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


This dataset is composed of 8 columns and 541909 rows. Let's eplore different columns ,check the null values and decide how we deal with them.

**Checking for null values**

In [15]:
df.isnull().values.any()

True

In [16]:
#Null values in each features
print("Null values in each features of the dataset")
df.isnull().sum()

Null values in each features of the dataset


InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [17]:
#Check the percentages of custemer missing 
print("Percentages of custemer missing :",round(df['CustomerID'].isnull().sum()*100/len(df),2),"%")

Percentages of custemer missing : 24.93 %


In [18]:
#Check the percentages of Desription missing
print("Percentages of Description missing :",round(df['Description'].isnull().sum()*100/len(df),2),"%")

Percentages of Description missing : 0.27 %


In [19]:
# Identify the number of NAs in each feature and select only those having NAs
total_NA = df.isnull().sum()[df.isnull().sum() != 0]

# Calculate the percentage of NA in each feature
percent_NA = df.isnull().sum()[df.isnull().sum() != 0]/df.shape[0]

# Summarize our findings in a dataframe
missing = pd.concat([total_NA, percent_NA], axis=1, keys=['Total NAs', 'Percentage']).sort_values('Total NAs', ascending=False)
missing

Total NAs  Percentage
CustomerID      135080    0.249267
Description       1454    0.002683

In [20]:
#Number of dulicate values in the dataset
print("Number of duplicate values :",df.duplicated(keep='first').sum())

Number of duplicate values : 5268


a. We have missing values in the CustomerID and Description columns.

b. Since 25% of the customer ID's are missing, we will create and fill a new column that has a 1 when customer ID is null and a 0 when it is not.

c. We will investigate the records where the CustomerID field is null to determine whether to erase those rows or fill in the missing values.

d. Since we won't be doing analysis on the descriptions of the orders, we can leave the null values as it is for now.

In [21]:
df['CustomerID_is_null'] = [1 if i == True else 0 for i in df.CustomerID.isnull()]

In [22]:
#Create a new dateframe of only orders with missing CustomerID values
df1 = df[df['CustomerID_is_null'] == 1]

In [23]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 135080 entries, 622 to 541540
Data columns (total 9 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   InvoiceNo           135080 non-null  object 
 1   StockCode           135080 non-null  object 
 2   Description         133626 non-null  object 
 3   Quantity            135080 non-null  int64  
 4   InvoiceDate         135080 non-null  object 
 5   UnitPrice           135080 non-null  float64
 6   CustomerID          0 non-null       float64
 7   Country             135080 non-null  object 
 8   CustomerID_is_null  135080 non-null  int64  
dtypes: float64(2), int64(2), object(5)
memory usage: 10.3+ MB


In [24]:
df1.head(10)

InvoiceNo StockCode                          Description  Quantity  \
622     536414     22139                                  NaN        56   
1443    536544     21773      DECORATIVE ROSE BATHROOM BOTTLE         1   
1444    536544     21774      DECORATIVE CATS BATHROOM BOTTLE         2   
1445    536544     21786                   POLKADOT RAIN HAT          4   
1446    536544     21787                RAIN PONCHO RETROSPOT         2   
1447    536544     21790                   VINTAGE SNAP CARDS         9   
1448    536544     21791   VINTAGE HEADS AND TAILS CARD GAME          2   
1449    536544     21801  CHRISTMAS TREE DECORATION WITH BELL        10   
1450    536544     21802      CHRISTMAS TREE HEART DECORATION         9   
1451    536544     21803       CHRISTMAS TREE STAR DECORATION        11   

        InvoiceDate  UnitPrice  CustomerID         Country  CustomerID_is_null  
622   12/1/10 11:52       0.00         NaN  United Kingdom                   1  
1443  12/1/10 14:32       2.51         NaN  United Kingdom                   1  
1444  12/1/10 14:32       2.51         NaN  United Kingdom                   1  
1445  12/1/10 14:32       0.85         NaN  United Kingdom                   1  
1446  12/1/10 14:32       1.66         NaN  United Kingdom                   1  
1447  12/1/10 14:32       1.66         NaN  United Kingdom                   1  
1448  12/1/10 14:32       2.51         NaN  United Kingdom                   1  
1449  12/1/10 14:32       0.43         NaN  United Kingdom                   1  
1450  12/1/10 14:32       0.43         NaN  United Kingdom                   1  
1451  12/1/10 14:32       0.43         NaN  United Kingdom                   1

I noticed that there are usually several products being bought per order where each unique InvoiceNo is a separate order.

In [25]:
print("Number of orders in df1:", df1['InvoiceNo'].nunique())
print("Avg. number of products per order in df1: ", round(df1['StockCode'].count() / df1['InvoiceNo'].nunique(),2) )
print("Avg. number of products per order in df: ",  round(df['StockCode'].count() / df['InvoiceNo'].nunique(),2) )

Number of orders in df1: 3710
Avg. number of products per order in df1:  36.41
Avg. number of products per order in df:  20.92


Since the customer ID's are missing, lets assume these orders were not made by the customers already in the data set because those customers already have ID's. We also don't want to assign these orders to those customers because this would alter the insights we draw from the data. Instead of dropping the null CustomerID values, let's assign those rows a unique customer ID per order. This will act as a new customer for each unique order.


**How should we assign the new customer ID's?**                                           




Using the values in the InvoiceNo column would be the most straightfoward approach. We will create a new customer ID column called NewID with the invoice numbers filling in for the missing values. Then we will add the number of unique orders in df1 and to number of unique values in CustomerID and see if it equals the number of unique values in NewID. This will check if any of the new values match the existing values in the column and make sure we didn't add more orders to an existing customer.

In [26]:
df['NewID'] = df['CustomerID']
df.loc[df['CustomerID_is_null'] == 1, ['NewID']] = df['InvoiceNo']

In [27]:
#drop CustomerID column since it isn't need anymore
#df = df.drop('CustomerID', 1)
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

    InvoiceDate  UnitPrice  CustomerID         Country  CustomerID_is_null  \
0  12/1/10 8:26       2.55     17850.0  United Kingdom                   0   
1  12/1/10 8:26       3.39     17850.0  United Kingdom                   0   
2  12/1/10 8:26       2.75     17850.0  United Kingdom                   0   
3  12/1/10 8:26       3.39     17850.0  United Kingdom                   0   
4  12/1/10 8:26       3.39     17850.0  United Kingdom                   0   

     NewID  
0  17850.0  
1  17850.0  
2  17850.0  
3  17850.0  
4  17850.0

Notice that NewID doesn't have any missing values, but the column includes object data types. As mentioned in the description of the dataset, if the invoice number starts with the letter "C", it means it the order was cancelled. All of the data types in the new customer ID column need to be numeric, so we need to first remove any "C" characters from the column.

In [28]:
#Remove all non digits from column NewID
df['NewID'] = df['NewID'].astype(str).str.replace('\D+', '')

#Convert to integer
df['NewID'] = pd.to_numeric(df['NewID'])

#Check if the 'C' character was removed in the NewID column
df[df['NewID']== 580605]

InvoiceNo  StockCode Description  Quantity    InvoiceDate  UnitPrice  \
524602   C580605  AMAZONFEE  AMAZON FEE        -1  12/5/11 11:36   17836.46   

        CustomerID         Country  CustomerID_is_null   NewID  
524602         NaN  United Kingdom                   1  580605

In [29]:
Num_CustomerID = df['CustomerID'].nunique()
Num_df1_InvoiceNo = df1['InvoiceNo'].nunique()
Num_NewID = df['NewID'].nunique()
print("Number of Customers:", df['CustomerID'].nunique())
print("Number of Orders in df1:", df1['InvoiceNo'].nunique())
print("Number of Customers + Number of Orders in df1:", Num_CustomerID + Num_df1_InvoiceNo)
print("Number of New ID's:", df['NewID'].nunique())

Number of Customers: 4372
Number of Orders in df1: 3710
Number of Customers + Number of Orders in df1: 8082
Number of New ID's: 8082


Since both values equal 8082, we know all the different orders that didn't have a customer ID got assigned unique NewID.